sui_core/src/authority/authority_checkpoints.rs

```diff
@@ -0,0 +1,1325 @@
+            new_locals.no_more_fragments = true;
+            self.set_locals(locals, new_locals)
+                .map_err(FragmentInternalError::Error)?;
+
+            // A little argument about how the fragment -> checkpoint process is live
+            //
+            // A global checkpoint candidate must contain at least 2f+1 stake. And as
+            // a result of this f+1 stake will be from honest nodes that by definition
+            // must have submitted a proposal (because it is included!).
+            // So f+1 honest authorities will be able to reconstruct and sign the
+            // checkpoint. And all other authorities by asking all authorities will be
+            // able to get f+1 signatures and construct a checkpoint certificate.
+
+            Err(FragmentInternalError::Error(
+                SuiError::GenericAuthorityError {
+                    error: "Missing info to construct known checkpoint.".to_string(),
+                },
+            ))
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Handles the submission of a full checkpoint externally, and stores
+    /// the certificate. It may be used to upload a certificate, or induce
+    /// the authority to catch up with the latest checkpoints.
+    ///
+    /// A cert without contents is only stored if we have already processed
+    /// internally the checkpoint. A cert with contents is processed as if
+    /// it came from the internal consensus.
+    pub fn handle_checkpoint_certificate(
+        &self,
+        checkpoint: &CertifiedCheckpoint,
+        contents: &Option<CheckpointContents>,
+    ) -> Result<CheckpointResponse, SuiError> {
+        // Get the record in our checkpoint database for this sequence number.
+        let current = self
+            .checkpoints
+            .get(&checkpoint.checkpoint.waypoint.sequence_number)?;
+
+        match &current {
+            // If cert exists, do nothing (idempotent)
+            Some(AuthenticatedCheckpoint::Certified(_current_cert)) => {}
+            // If no such checkpoint is known, then return an error
+            // NOTE: a checkpoint must first be confirmed internally before an external
+            // certificate is registered.
+            None => {
+                if let &Some(contents) = &contents {
+                    // Check and process contents
+                    checkpoint.check_transactions(&self.committee, contents)?;
+                    self.handle_internal_set_checkpoint(checkpoint.checkpoint.clone(), contents)?;
+                    // Then insert it
+                    self.checkpoints.insert(
+                        &checkpoint.checkpoint.waypoint.sequence_number,
+                        &AuthenticatedCheckpoint::Certified(checkpoint.clone()),
+                    )?;
+
+                    // Now that we have the new checkpoint we try to move forward the checkpoint creation
+                    // process
+                    loop {
```

Comment:
Just to clarify, and seek your advice here: the intent of this is to make progress until a checkpoint is complete, and then check if the data for any subsequent checkpoint is already available and can be used to make more checkpoints. It does not wait on external events or more info to arrive. Do you feel that I should refactor (any advice very welcome here).

Reply:
I understand you're not waiting for any data, and that this is push-forward loop. But it will come at:
- at a latency cost (now any specific handler can take a while to complete),
- a concurrency hardening cost (several busy loops may hit the same backing data store concurrently),
- a code complexity cost (several places where this busy loop lives).

You could give such a push-forward logic one single place to live, and then notify it from the various handlers you deal with using a range of means ranging from a [condvar](https://doc.rust-lang.org/std/sync/struct.Condvar.html) to a. good old channel message, to an [observer pattern](https://github.com/rxRust/rxRust).

Reply:
Yep included in follow-up issue